File: src/merkle/getMerkleRootFromProof.ts
import demosMemory from "./memory";

import libdemos from "@libdemos";

import { crypto_hash_sha512_BYTES } from "../utils/interfaces";

/**
 * @function
 * Calculates the Merkle root from the element hash and its Merkle proof.
 *
 * @param hash: The hash of the base element in question.
 * @param proof: The first element is the first leave that was added for the calculation etc. The last
 * byte is either 0 or 1, indicating whether it is to the left or to the right in the tree.
 *
 * @returns The Merkle root
 */
const getMerkleRootFromProof = async (
  hash: Uint8Array,
  proof: Uint8Array,
): Promise<Uint8Array> => {
  const proofLen = proof.length;
  if (proofLen % (crypto_hash_sha512_BYTES + 1) !== 0)
    throw new Error("Proof length not multiple of hash length + 1.");
  const proofArtifactsLen = proofLen / (crypto_hash_sha512_BYTES + 1);

  const wasmMemory = demosMemory.verifyMerkleProofMemory(proofLen);
  const demosModule = await libdemos({
    wasmMemory,
  });

  const ptr1 = demosModule._malloc(crypto_hash_sha512_BYTES);
  const elementHash = new Uint8Array(
    wasmMemory.buffer,
    ptr1,
    crypto_hash_sha512_BYTES,
  );
  elementHash.set(hash);

  const ptr2 = demosModule._malloc(proofLen);
  const proofArray = new Uint8Array(wasmMemory.buffer, ptr2, proofLen);
  proofArray.set(proof);

  const ptr3 = demosModule._malloc(crypto_hash_sha512_BYTES);
  const rootArray = new Uint8Array(
    wasmMemory.buffer,
    ptr3,
    crypto_hash_sha512_BYTES,
  );

  const result = demosModule._get_merkle_root_from_proof(
    proofArtifactsLen,
    elementHash.byteOffset,
    proofArray.byteOffset,
    rootArray.byteOffset,
  );

  demosModule._free(ptr1);
  demosModule._free(ptr2);

  switch (result) {
    case 0: {
      const proof = Uint8Array.from(rootArray);
      demosModule._free(ptr3);

      return proof;
    }

    case -1: {
      demosModule._free(ptr3);

      throw new Error(
        "Could not allocate memory for hash concatenation helper array.",
      );
    }

    case -2: {
      demosModule._free(ptr3);

      throw new Error("Proof artifact position is neither left nor right.");
    }

    case -3: {
      demosModule._free(ptr3);

      throw new Error("Could not calculate hash.");
    }

    default: {
      demosModule._free(ptr3);

      throw new Error("Unexpected error occured.");
    }
  }
};

export default getMerkleRootFromProof;

File: src/merkle/get_merkle_root_from_proof.c
#include <stdbool.h>
#include <stdint.h>
#include <string.h>

#include "../../libsodium/src/libsodium/include/sodium/crypto_hash_sha512.h"

int
get_merkle_root_from_proof(
    const unsigned int PROOF_ARTIFACTS_LEN,
    const uint8_t element_hash[crypto_hash_sha512_BYTES],
    const uint8_t proof[PROOF_ARTIFACTS_LEN * (crypto_hash_sha512_BYTES + 1)],
    uint8_t root[crypto_hash_sha512_BYTES])
{
  memcpy(root, element_hash, crypto_hash_sha512_BYTES);

  size_t i;
  unsigned int position;
  int res;

  if (PROOF_ARTIFACTS_LEN == 1)
  {
    res = memcmp(proof, element_hash, crypto_hash_sha512_BYTES);
    if (res == 0)
    {
      memcpy(root, element_hash, crypto_hash_sha512_BYTES);

      return 0;
    }
    else
    {
      // If 1 artifact then either proof is the same
      // as the element hash, and therefore
      // the same as the root, or we have an error
      return -1;
    }
  }

  uint8_t(*concat_hashes)[crypto_hash_sha512_BYTES]
      = malloc(sizeof(uint8_t[2][crypto_hash_sha512_BYTES]));
  if (concat_hashes == NULL) return -1;

  for (i = 0; i < PROOF_ARTIFACTS_LEN; i++)
  {
    position
        = proof[i * (crypto_hash_sha512_BYTES + 1) + crypto_hash_sha512_BYTES];

    // Proof artifact goes to the left
    if (position == 0)
    {
      memcpy(concat_hashes[0], &proof[i * (crypto_hash_sha512_BYTES + 1)],
             crypto_hash_sha512_BYTES);
      memcpy(concat_hashes[1], root, crypto_hash_sha512_BYTES);
    }
    else if (position == 1)
    {
      memcpy(concat_hashes[0], root, crypto_hash_sha512_BYTES);
      memcpy(concat_hashes[1], &proof[i * (crypto_hash_sha512_BYTES + 1)],
             crypto_hash_sha512_BYTES);
    }
    else
    {
      free(concat_hashes);

      return -2;
    }

    res = crypto_hash_sha512(root, concat_hashes[0],
                             2 * crypto_hash_sha512_BYTES);
    if (res != 0)
    {
      free(concat_hashes);

      return -3;
    }
  }

  free(concat_hashes);

  return 0;
}
